### ISSUE TITLE
Improve how transaction states are processed and shown to users, by updating the Tx Controller

### ISSUE THREAD

**Issue by estebanmino** [MetaMask/metamask-mobile#2271]
## Epic info
This work represents the [Reliable transaction state Epic](https://www.notion.so/Reliable-transaction-state-5dea56ff7b4946cd9a3160c3e5e3a6f7)

Improve how we process transaction state better by updating a txn state controller. The extension currently has a **transaction state controller** to handle user scenarios, something we lack.

## Use cases
Cases where our transaction receipts are failing
- On transaction history we're currently using the gas values that the user sent the transaction with, but in many cases the actual gas spent was less so we are showing misleading informaction.
- Wrong transaction state: showing failed transaction, when the tx was successful.
- Wrong transaction state: “Dropped and replaced” shows up in etherscan and our end shows pending.
- Wrong value being shown on history: showing gas fee instead of the amount swapped. 
- The states of transactions 2 transactions with same nonce are shown wrongly.

## Proposed solutions

- [x] Etherscan reconciler to address Mainnet and Testnet issues.
- [x] Blockchain reconciler to address custom networks issues (BSC, Polygon, etc.)
- [x] Transaction stuck in "submitted" state.
- [ ] Transaction issue regarding `nonce` - `intendId`.

## Impact of this work
The intent of this work is to build a controller that works for Mobile, and then works for the Extension product as the team moves away from its own controller

## Stop-gap Solution 
Implemented for EIP-1559 (on the UI side) as part of v3.0: https://github.com/MetaMask/metamask-mobile/pull/2909

**Comment by omnat** [comment]
This issue was found while QAing the PR for custom nonce feature. Related to same issue of not having logic to capture transaction status. 

![image.png](https://images.zenhubusercontent.com/5c18e51bb423df7568ac3781/46f47a52-8090-4182-8afa-2afb4cc9234b)

**Comment by omnat** [comment]
Related #2382

**Comment by mobularay** [comment]
Extension has a their own receipts controller, but this is not a shared controller.

**Comment by mobularay** [comment]
Dev team to create own controller (based on Extension's work) to resolve these cases; need to revisit once both teams are ready to develop a cross-platform controller

For reference, Extension's list of controllers: https://github.com/MetaMask/metamask-extension/tree/develop/app/scripts/controllers/transactions

The controller we write should be in new controller design, which is something extension will migrate to in future. A sync with @Gudahtt to be done as first step on implementing this ticket.

**Comment by mobularay** [comment]
More details provided by @omnat on Slack, for any Engineer that is picking up this work: https://consensys.slack.com/archives/GBW7S9FSN/p1620089921461100?thread_ts=1620081083.460800&cid=GBW7S9FSN

1. It’s a big one, and I think a call would be an efficient way to kick-start this one.
2. to get more context from @estebanmino 
3. also looping in @brad-decker  would be good. Brad has been working on the transaction controller lately - I don’t know if tx state controller was part of your focus 
4. Since its about writing a new controller that mobile doesn’t have so far, we should write it with the new controller design in mind - so I’d suggest inviting @Gudahtt to this call


As per item 1 above, the first Engineer that is picking up this work should understand what work needs to be done, and break this "big" story down into smaller work pieces so that multiple Engineers can work on them in parallel

**Comment by sethkfman** [comment]
Discussed with @estebanmino @brad-decker @Gudahtt and it was agreed that a new txs state controller should be built in the shared repo with design and concepts taken from what the extension has done. This new controller will be designed to find a common state that both the extension and mobile app can use. It will also be built to address problems on both platforms.

- Ref. https://github.com/MetaMask/metamask-extension/blob/09d81ac5f2c94d7730d2e981b4f05734858030bf/shared/constants/transaction.js

**Comment by mobularay** [comment]
@sethkfman has been gathering additional information, and will break-down this card into smaller chunks of work.

**Comment by omnat** [comment]
This part needs to be divided further @sethkfman to work with @estebanmino for this. Indicate what we can parallelize. 

Prioritization: Important as it leads to improved transaction reliability (primary experience of MetaMask)

Milestones / Process:
1. Design the transaction state controller for shared controller repo. This technical proposal needs to be reviewed by extension-mobile teams. 
2. Mobile app will migrate to this controller to improve how transaction states are represented, how nonce management works, the speed up/cancel actions (and its reliability)
3. The UI changes that are required are on the naming of the state change

Considerations from experience of swaps controller
1. Write tests to test it both on extension & mobile
2. Code interface (fill in the gaps) 

Out of Scope - 
1. Migration of extension controller to this shared controller (extension team will do this in their timeline)
2. UI update of transaction history???
